alf/utils/video_recorder.py

```diff
@@ -0,0 +1,323 @@
+    """
+    if width is not None and height is not None:
+        return cv2.resize(img, dsize=(width, height))
+    elif width is not None:
+        scale = float(width) / img.shape[1]
+    elif height is not None:
+        scale = float(height) / img.shape[0]
+    else:
+        raise ValueError('At least width or height should be provided.')
+    return cv2.resize(
+        img, dsize=(0, 0), fx=scale, fy=scale, interpolation=interpolation)
+
+
+def _generate_img_matrix(imgs, cols, img_width):
+    """Given a list of images, arrange them in a image matrix that has ``cols``
+    columns and a width of ``img_width``. Each cell corresponds to an image in
+    the list.
+
+    Args:
+        imgs (list[np.array]): a list of np.array images
+        cols (int): number of columns
+        img_width (int): the output image width
+
+    Returns:
+        np.array: the output image, with redundant cells as 0s
+    """
+    W = img_width // cols
+    rows = (len(imgs) + cols - 1) // cols
+    imgs = [_resize_image(i, width=W) for i in imgs]
+    H = max([i.shape[0] for i in imgs])
+
+    ret_img = np.ones((H * rows, W * cols, 3), dtype=np.uint8) * 255
+    for i, im in enumerate(imgs):
+        r, c = i // cols, i % cols
+        ret_img[r * H:r * H + im.shape[0], c * W:(c + 1) * W, :] = im
+    return ret_img
+
+
+@gin.configurable(
+    whitelist=['img_plot_width', 'value_range', 'frames_per_sec'])
+class VideoRecorder(GymVideoRecorder):
+    """A video recorder that supports plotting prediciton info in addition to
+    rendering frames.
+
+    Suppors the following data types:
+
+    * action distribution: plotted as a probability curve. For a discrete
+    distribution, the probabilities are directly plotted. For a continuous
+    distribution, the curve will be approximated by sampled points for each
+    action dimension.
+    * action: plotted as heatmaps. Long type actions will be first
+    converted to one-hot encodings.
+
+    ``env_frame`` is a ``numpy.ndarray`` with shape ``(x, y, 3)``,
+    representing RGB values for an x-by-y pixel image, output from
+    ``env.render('rgb_array')``.
+    """
+
+    def __init__(self,
+                 env,
+                 img_plot_width=640,
+                 value_range=1.,
+                 frames_per_sec=30,
+                 **kwargs):
+        """
+        Args:
+            env (Gym.env):
+            img_plot_width (int): the image width for displaying prediction info,
+                excluding the env frame width
+            value_range (float): if quantities plotted as heatmaps, the values
+                will be clipped according to ``[-value_range, value_range]``.
+            frames_per_sec (fps):
+        """
+        super(VideoRecorder, self).__init__(env=env, **kwargs)
+        self._img_plot_width = img_plot_width
+        self._dist_imgs_per_row = 4
+        self._value_range = value_range
+        self.frames_per_sec = frames_per_sec  # overwrite the base class
```

Comment:
frames_per_sec should be retrieved from env.metadata['video.frames_per_second']

Reply:
As I just verified, we can overwrite the fps to render a slow-motion video by specifying a smaller fps. This might be convenient to analyze the action outputs.